msal/src/main/java/com/microsoft/identity/client/AccountAdapter.java

```diff
@@ -1,125 +1,311 @@
 import android.support.annotation.Nullable;
 
+import com.microsoft.identity.common.internal.cache.ICacheRecord;
 import com.microsoft.identity.common.internal.dto.AccountRecord;
-import com.microsoft.identity.common.internal.dto.IAccountRecord;
-import com.microsoft.identity.common.internal.logging.Logger;
-import com.microsoft.identity.common.internal.providers.microsoft.MicrosoftAccount;
 import com.microsoft.identity.common.internal.providers.oauth2.OAuth2TokenCache;
 import com.microsoft.identity.common.internal.util.StringUtil;
 
-/**
- * Adapter class for Account transformations.
- */
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
 class AccountAdapter {
 
     private static final String TAG = AccountAdapter.class.getSimpleName();
 
     /**
-     * Adapts the {@link AccountRecord} type to the MSAL-exposed {@link IAccount}.
-     *
-     * @param accountIn The Account to transform.
-     * @return A representation of the supplied Account, as an IAccount.
+     * Abstract class representing a filter for ICacheRecords.
      */
-    @NonNull
-    static IAccount adapt(@NonNull final IAccountRecord accountIn) {
-        final String methodName = ":adapt";
-        final com.microsoft.identity.client.Account accountOut
-                = new com.microsoft.identity.client.Account();
-
-        // Populate fields...
-        final IAccountIdentifier accountId;
-        final IAccountIdentifier homeAccountId;
-
-        if (MicrosoftAccount.AUTHORITY_TYPE_V1_V2.equals(accountIn.getAuthorityType())) {
-            Logger.info(
-                    TAG + methodName,
-                    "Account type is AAD"
-            );
+    private interface CacheRecordFilter {
 
-            // This account came from AAD
-            homeAccountId = new AzureActiveDirectoryAccountIdentifier() {{ // This is the home_account_id
-                // Grab the homeAccountId
-                final String homeAccountIdStr = accountIn.getHomeAccountId();
+        List<ICacheRecord> filter(@NonNull final List<ICacheRecord> records);
+    }
 
-                // Split it into its constituent pieces <uid>.<utid>
-                final String[] components = homeAccountIdStr.split("\\.");
+    /**
+     * A filter for ICacheRecords that filters out home or guest accounts, based on its
+     * constructor initialization.
+     */
+    private static class HomeAccountFilter implements CacheRecordFilter {
 
-                // Set the full string value as the identifier
-                setIdentifier(homeAccountIdStr);
+        private final boolean mInverseMatch;
 
-                // Set the uid as the objectId
-                setObjectIdentifier(components[0]);
+        /**
+         * Creates a new HomeAccountFilter.
+         *
+         * @param inverseMatch If true, only guest accounts will be returned. If false, only home.
+         */
+        HomeAccountFilter(final boolean inverseMatch) {
```

Comment:
Also for better readability you can probably use Enum 'AccountTypeToFilter' with values Home and  Guest.

Reply:
Since you've been using polymorphism for other filters.... although it's convenient in this case to just "switch" the mode.  I'd suggest sticking with polymorphism and implementing a GuestAccountFilter.

Reply:
Hah - I actually _had_ a guest account filter and then refactored it out. Easy to replace 👍

Reply:
Resolved in ad770bbde46de73922851e1ec8151d7152d2ffcf